Traduora caching

vullnetyy opened the pull request:
I've implemented a configurable timer refresh for calling traduora periodically. 

In case the dictionary is null, I chose to synchronously fetch the translations instead of providing some empty dictionary. I haven't really thought this through, I just went for it so we can have a discussion afterward.

Please let me know what you think.

vullnetyy committed:
Implemented StringLocalizer synchronously

vullnetyy committed:
Created and used a cache for Traduora data

vullnetyy committed:
Fixed HttpClient mutation

Moved service config to ServiceExtensions

vullnetyy committed:
Added a config based refresh timer function to CacheManager

vullnetyy committed:
Merge branch 'master' into traduora-caching

perokvist commented on sample/Web/Controllers/HomeController.cs:
```diff
@@ -1,42 +1,30 @@
+        private IStringLocalizer _localizer;
 
-        public HomeController(IConfiguration config, IHttpClientFactory httpClientFactory)
+        public HomeController(IConfiguration config, IStringLocalizer localizer, IHttpClientFactory httpClientFactory)
```
Unused HttpClientFactory.

perokvist replied:
Uused config.

perokvist commented on sample/Web/ServiceExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+            services.AddHttpClient(HttpClientName)
+                .ConfigureHttpClient((sp, client) =>
+            {
+                var config = sp.GetRequiredService<IConfiguration>();
```
Next step - try adding typed config

perokvist commented on sample/Web/ServiceExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+
+            services.AddSingleton<IStringLocalizer>(sp =>
+            {
+                var config = sp.GetRequiredService<IConfiguration>();
```
When both Config and Service is in the IoC container. The need for resolving them in registration is not needed.

perokvist commented on sample/Web/ServiceExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+
+                var cache = new CacheManager(sp.GetRequiredService<ITraduoraService>().GetTranslations, refreshMilliseconds);
+
+                return new StringLocalizer(cache.DataProvider);
```
If the provider also I registered.

perokvist commented on sample/Web/TraduoraService.cs:
```diff
@@ -0,0 +1,43 @@
+            //Change the 3 variables below for your Traduora project and API Key=======
+            string clientId = _config["Traduora2:ClientId"];
+            string clientSecret = _config["Traduora2:Secret"];
+            string projectId = "03aac4d9-a898-49f0-8546-1343c2964b4a";
```
Move to config

perokvist commented on sample/Web/TraduoraService.cs:
```diff
@@ -0,0 +1,43 @@
+        private readonly IConfiguration _config;
+        private readonly IHttpClientFactory _httpClientFactory;
+
+        public TraduoraService(IConfiguration config, IHttpClientFactory httpClientFactory)
```
Possible to take a TrudoraClientFactory here

perokvist commented on sample/Web/TraduoraService.cs:
```diff
@@ -0,0 +1,43 @@
+
+            HttpClient httpClient = _httpClientFactory.CreateClient(ServiceExtensions.HttpClientName);
+
+            var traduoraClient = new TraduoraClient(httpClient);
```
Invoke factory ?

perokvist commented on src/Core/CacheManager.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            if (_data != null) return _data;
+
+            _data = _asyncDataProvider().Result;
```
Next step, do this async.
How to Handle startup. https://andrewlock.net/running-async-tasks-on-app-startup-in-asp-net-core-part-1/
Kika med Eric om alternativ och tankar från nuvarande Localizer i Assets.

vullnetyy committed:
Injecting TraduoraClient from DI

vullnetyy committed:
Renamed to Traduora.Provider and moved projectId to config

allrameest commented on sample/Web/Controllers/HomeController.cs:
```diff
@@ -1,42 +1,24 @@
-            JObject exportedJson = await traduoraClient.GetTranslations(projectId, locale, key);
-
-            return $"This is the home controller speaking. \n\nHere are some translations: \n\n{exportedJson}";
+            return View();
```
An example with IViewLocalizer used in the view would be nice!

vullnetyy replied:
Absolutely. I'm on it!

allrameest commented on sample/Web/ServiceExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+using Web;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.Extensions.DependencyInjection
```
Seems like this could just be placed in the Web namespace.

allrameest commented on src/Core/CacheManager.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Core
+{
+    public class CacheManager
```
I'm thinking this could be named better to say what it is. And maybe also have a more generic implementation. Like RefreshingValue or something like that.

And signature for it could be:
```csharp
class RefreshingValue<T>
    ctor(Func<Task<T>> valueFactory)
    T Value { get; }
```

vullnetyy replied:
I followed your suggestion and created something called AsyncPollingCache<T>. I also provide a DictionaryPollingCache : AsyncPollingCache<IReadOnlyDictiona........> as an easy default to use for our current traduora use case.

Do you think it's better now?

allrameest commented on src/Core/CacheManager.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public Func<IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>> DataProvider => GetData;
+
+        public CacheManager(Func<Task<IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>>> asyncDataProvider, int refreshMilliseconds)
```
Constructor should be moved up. As a rule order should be fields, constructor, properties, methods. With exceptions of course...

allrameest replied:
I would suggest `TimeSpan refreshInterval` instead of `int refreshMilliseconds`.

allrameest commented on src/Core/StringLocalizer.cs:
```diff
@@ -1,31 +1,42 @@
-    public class StringLocalizer : IStringLocalizer
-    {
-        readonly Func<IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>> modelProvider;
+	public class StringLocalizer : IStringLocalizer
```
Would be good to name this for what it used for localization. E.g. DictionaryStringLocalizer could work.

allrameest commented on src/Core/StringLocalizer.cs:
```diff
@@ -1,31 +1,42 @@
+using System.Linq;
+using Microsoft.Extensions.Localization;
 
 namespace Core
```
Could this project maybe be named Localizer or Localization?

allrameest commented on src/Core/StringLocalizer.cs:
```diff
@@ -1,31 +1,42 @@
+        private CultureInfo _culture;
+        private CultureInfo Culture => _culture ?? CultureInfo.CurrentUICulture;
 
         public StringLocalizer(Func<IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>> modelProvider)
```
Since StringLocalizer gets the culture in the constructor, you can lookup the dictionary for the culture and store that in a field. Then all methods can use that field and can skip one dictionary lookup.

allrameest replied:
Of course that could be a problem if StringLocalizer is a singleton per culture. Not sure about the lifetime of those...

vullnetyy replied:
I don't really think we should do that. I just created that second constructor for convenience, expecting that we would need it in the future. 

In an effort to give control of fetching/caching of the strings to another object, the string localizer always needs to look up the string using the _modelProvider for the relevant culture. Thus, it is the _modelProvider's responsibility to be efficient about it.

allrameest commented on src/Traduora/TraduoraClient.cs:
```diff
@@ -1,38 +1,70 @@
-            var traduoraApi = RestService.For<ITraduoraApi>(_httpClient);
+            var output = new Dictionary<string, IReadOnlyDictionary<string, string>>();
+
+            JObject localesJson = await _traduoraApi.GetLocales(projectId, $"Bearer {authToken}");
```
Create a model that can be deserialized to instead of JObject/JToken/JWhatever. Would make code simpler and typed.

allrameest commented on sample/Web/ServiceExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+{
+    public static class ServiceExtensions
+    {
+        public const string HttpClientName = "traduora";
```
This could be private.

allrameest commented on src/Traduora/TraduoraClient.cs:
```diff
@@ -1,38 +1,70 @@
             string format = DefaultFormat)
         {
-            var traduoraApi = RestService.For<ITraduoraApi>(_httpClient);
+            var output = new Dictionary<string, IReadOnlyDictionary<string, string>>();
```
Pass StringComparer.OrdinalIgnoreCase to make it case insensitive.

allrameest commented on src/Traduora/TraduoraClient.cs:
```diff
@@ -1,38 +1,70 @@
-            JToken responseJson = await traduoraApi.GetToken(authBody);
+        private static IReadOnlyDictionary<string, string> ToReadOnlyDictionary(JObject exportedJson)
+        {
+            var output = new Dictionary<string, string>();
```
Pass StringComparer.OrdinalIgnoreCase to make it case insensitive.

vullnetyy committed:
Changed small issues related to PR

Changed namespace of ServiceExtensions
Changed int to Timespan for refreshMilliseconds at CacheManager
Ordered the members of CacheManager correctly
Renamed the Core project and StringLocalizer class
Deleted StringLocalizerFactory because it was unused

vullnetyy committed:
Refactored CacheManager into a generic AsyncPollingCache

vullnetyy committed:
Created strongly typed configurations and added basic config validation

perokvist commented:
Checkout the Options pattern within asp.net core, and use that.
This is what CMP uses and most 3rd party libraries.
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-2.2

It supports validation as well.

vullnetyy commented:
This is using Options under the hood. It just hides the options from you so
you don’t have to deal with that extra layer of indirection.

vullnetyy commented:
https://www.google.com/amp/s/andrewlock.net/adding-validation-to-strongly-typed-configuration-objects-in-asp-net-core/amp/

By the way, did you mean that I should use attributes for validation
instead?

perokvist commented:
Please use "standard" options for configuration, that what we use within CMP.
Move TraduoraService to lib (interface not need, no one else will impl?)
Settings should also be moved to lib.
Move Concreate InMemory cache and interface to Core. (remove cache lib)
Focus on solving startup and async fetch.
The current "facade" for the provider the traduora service only supports one project.
Keep it simple :)

vullnetyy commented:
Let’s talk about all these details tomorrow. I’ll then make a list and execute

vullnetyy commented:
@perokvist @allrameest I'm closing this pull request and I will make a new one that will include the following changes:

* Async main method that initializes translation dictionary
* Strongly typed JSON deserialization
* A clear separation of concerns on the Localization library, Traduora.Provider library, Traduora.Localization library (that uses the first two) and then a Web library that uses Traduora.Localization in a very simple way
* An example with IViewLocalizer (taking advantage of IStringLocalizer in the background)

vullnetyy committed:
Used IOptions and removed the Nefit options nuget
